### PR TITLE
Cleanup crowbar-backup, and make restoration work

### PR DESCRIPTION
**Pull request by vuntz**

### CODE REVIEW

File: backup/crowbar-backup [hunk 1]
```diff
@@ -226,54 +272,90 @@ function restore()
...
+    exit 84
   fi
   echo
 
+  rm /tmp/chef-ready
+
   # Make sure install finished correctly
   if ! [ -e /opt/dell/crowbar_framework/.crowbar-installed-ok ]; then
-    echo "Crowbar \".crowbar-installed-ok\" marker missing"
+    echo "install-suse-cloud didn't successfully complete"
```
Comment (dirkmueller) [inline backup/crowbar-backup]: either "complete successfully" or "has failed."

File: backup/crowbar-backup [hunk 2]
```diff
@@ -226,54 +272,90 @@ function restore()
...
-    ## copied from install-suse-cloud
-    rm -rf /opt/dell/crowbar_framework/db/migrate
-    rm -f /opt/dell/crowbar_framework/db/{production.sqlite3,schema.rb}
-    raketasks="db:sessions:create $raketasks"
+  # Remember barclamps that we have locally, that were not in the backup
+  reinstall_barclamps=
+  if [ -f /var/lib/crowbar/backup/barclamp-list ]; then
+    for i in /opt/dell/crowbar_framework/barclamps/*.yml; do
+      barclamp=$(basename $i | sed "s/\.yml//g")
+      grep -q "^$barclamp$" /var/lib/crowbar/backup/barclamp-list && continue
```
Comment (dirkmueller) [inline backup/crowbar-backup]: grep -qx

File: backup/crowbar-backup [hunk 3]
```diff
@@ -210,7 +213,50 @@ function restore()
...
+      echo "   $missing_barclamps"
+      exit 83
+    fi
+  fi
+
+  restore_files etc/crowbar
+  restore_files $ETCFILES
+
+  # set the correct hostname
+  RESTORE_HOSTNAME=$(cat /etc/HOSTNAME)
```
Comment (dirkmueller) [inline backup/crowbar-backup]: useless use of cat award!

Comment (tboerger) [inline backup/crowbar-backup]: Why?

**Comment by dirkmueller** [comment]
Generally I'm fine with the changes, although it feels to me that some of the commits (at least those fixing up an earlier commit) should be squashed together. The fine granularity makes reviewing easier though, so thanks for that. My feeling is that some of the cleanups are subtle and have a certain risk of breaking something last minute. 

+1

File: backup/crowbar-backup (hunk 2 above)
Comment (vuntz) [inline backup/crowbar-backup]: Oh, didn't know `-x`. Thanks for the hint!

File: backup/crowbar-backup (hunk 3 above)
Comment (dirkmueller) [inline backup/crowbar-backup]: The purpose of cat is to concatenate (or "catenate") files. If it's only one file, concatenating it with nothing at all is a waste of time, and costs you a process. See http://www.smallo.ruhr.de/award.html#cat

In general the times were we optimized runtimes of shell scripts are long over.. hence the smiley.

File: backup/crowbar-backup (hunk 1 above)
Comment (vuntz) [inline backup/crowbar-backup]: fixed

File: backup/crowbar-backup (hunk 3 above)
Comment (vuntz) [inline backup/crowbar-backup]: I'm actually curious how to assign the content of a file to a variable without `cat`, in a way that is not bash-specific :-)

Comment (aspiers) [inline backup/crowbar-backup]: `foo=$(<file)` works in `zsh` too (in fact I suspect `bash` stole the idea from `zsh` as per usual) so technically it's not specific to just `bash` ;-)

Comment (dirkmueller) [inline backup/crowbar-backup]: The script uses #!/bin/bash so I don't see an issue with that..

Comment (aspiers) [inline backup/crowbar-backup]: Agreed - it already uses other non-POSIX features such as `[[ ]]` anyway.

**Comment by vuntz** [comment]
This has survived quite some testing.

**Pull request merged**